src/coreComponents/physicsSolvers/contact/SolidMechanicsAugmentedLagrangianContact.cpp

```diff
@@ -0,0 +1,1347 @@
+
+    quadList.resize( nQuad );
+    triList.resize( nTri );
+    arrayView1d< localIndex > const quadList_v = quadList.toView();
+    arrayView1d< localIndex > const triList_v = triList.toView();
+
+    forAll< parallelDevicePolicy<> >( nTri, [triList_v, vals_v] GEOS_HOST_DEVICE ( localIndex const kfe )
+    {
+      triList_v[kfe] = vals_v[kfe];
+    } );
+
+    forAll< parallelDevicePolicy<> >( nQuad, [quadList_v, vals_v, nTri] GEOS_HOST_DEVICE ( localIndex const kfe )
+    {
+      quadList_v[kfe] = vals_v[nTri+kfe];
+    } );
+
+    this->m_faceTypesToFaceElements[meshName]["Quadrilateral"] =  quadList;
+    this->m_faceTypesToFaceElements[meshName]["Triangle"] =  triList;
+  } );
+
+}
+
+void SolidMechanicsAugmentedLagrangianContact::createBubbleCellList( DomainPartition & domain ) const
+{
+
+  forDiscretizationOnMeshTargets( domain.getMeshBodies(), [&] ( string const &,
+                                                                MeshLevel & mesh,
+                                                                arrayView1d< string const > const regionNames )
+  {
+    ElementRegionManager & elemManager = mesh.getElemManager();
+
+    SurfaceElementRegion const & region = elemManager.getRegion< SurfaceElementRegion >( getUniqueFractureRegionName() );
+    FaceElementSubRegion const & subRegion = region.getUniqueSubRegion< FaceElementSubRegion >();
+    array1d< localIndex > tmpSpace( 2*subRegion.size());
+    SortedArray< localIndex > faceIdList;
+
+    arrayView1d< localIndex > const tmpSpace_v = tmpSpace.toView();
+    // Store indexes of faces in the temporany array.
+    {
+      ArrayOfArraysView< localIndex const > const elemsToFaces = subRegion.faceList().toViewConst();
+
+      forAll< parallelDevicePolicy<> >( subRegion.size(), [ tmpSpace_v, elemsToFaces ] GEOS_HOST_DEVICE ( localIndex const kfe )
+      {
+
+        localIndex const kf0 = elemsToFaces[kfe][0], kf1 = elemsToFaces[kfe][1];
+        tmpSpace_v[2*kfe] = kf0, tmpSpace_v[2*kfe+1] = kf1;
+
+      } );
+    }
+
+    // Sort indexes to enable efficient searching using binary search.
+    RAJA::stable_sort< parallelDevicePolicy<> >( tmpSpace_v );
+    faceIdList.insert( tmpSpace_v.begin(), tmpSpace_v.end());
+
+    // Search for bubble element on each CellElementSubRegion and
+    // store element indexes, global and local face indexes.
+    elemManager.forElementSubRegions< CellElementSubRegion >( regionNames, [&]( localIndex const, CellElementSubRegion & cellElementSubRegion )
+    {
+
+      arrayView2d< localIndex const > const elemsToFaces = cellElementSubRegion.faceList().toViewConst();
+
+      RAJA::ReduceSum< ReducePolicy< parallelDevicePolicy<> >, localIndex > nBubElems_r( 0 );
+
+      localIndex const n_max = cellElementSubRegion.size() * elemsToFaces.size( 1 );
+      array1d< localIndex > keys( n_max );
+      array1d< localIndex > perms( n_max );
+      array1d< localIndex > vals( n_max );
+      array1d< localIndex > localFaceIds( n_max );
```

Comment:
if you only need these in here and they don't need to go on device I would use `std::vector`

Reply:
These variables are necessary to compute the lists of elements and face required to assemble the bubble contributions.
All the operations to compute these lists are designed to run on GPU, so yes, they need to go on device.